### PR TITLE
Feature - Todo list

### PR DESCRIPTION
**Pull request by asherbig**
Hey man, it's been a pleasure working on this site. I hope my code is up-to-par. I tried my best to optimize things that could potentially cause problems, and I also made an effort not to copy/paste code where it was easy.

Refactor spots: I already have a few places that I know I want to go back and refactor now that I know how they should be structured, but they are strictly code-cleanliness things that bother me. 
* The rewards filter service. (it's a mess of maps that serve as temporary lookups for functions that are chained, but still have to be called in an order to make sure everything works. Not proud of that service at all. comments have been left there.
* The filter panels. Their UI is very copy-paste. They share a ton of the same code and could definitely be broken into a generic component.

Additional features: I still have things that I want to add to this. I feel like with any good side project, the possibilities for improvement or more features is never-ending. When this is merged, I'd like to make issues on your repo that I'll work on and push out as updates.
* Refresh button
* Auto-refresh
* Column re-organization on refresh (maybe? similar to how DIM does it. I feel like people are familiar with that pattern)
* Compact view (fitting more rows on the page would be a big plus)
* Fix double scroll-bar (this annoys me to no end) - Fixed in this PR with a workaround. Not taking advantage of AG grid's viewport rendering, but there's not enough rows to make a noticeable performance difference.
* More detailed bounty progress
* Don't re-use d2c-milestone-check. (ideally create a generic progress component that works for bounties and milestones, and takes a generic progress model)
* Bounty expiration feed (similar to notification style popups)
* Save and load filter/sort states
* Fix spider bounties issue with filters not recognizing them as the same as the previous week (hard)
* Custom sort order (look into drag & drop)
* Overlay tier info on gear engram icons
* Better quick filters or quick filter placement (I want to get community input for suggestions)

### CODE REVIEW

File: src/_util.scss [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  margin-bottom: 10px;
+}
+
+.margin-bottom-0 {
+  margin-bottom: 0 !important;
+}
+
+.margin-left-10 {
+  margin-left: 10px !important;
+}
```
Comment (asherbig) [inline src/_util.scss:19]: The company I work for has stylesheets that give utilities, layouts, and other commonly needed css. I think it helps reduce the amount of custom css classes and the need to put margin-top: x in all these different component scss files as part of some custom class definition.

File: src/ag-grid-styles.scss [hunk 2]
```diff
@@ -0,0 +1,30 @@
+
+@import "./node_modules/@ag-grid-community/core/dist/styles/ag-grid.scss";
```
Comment (asherbig) [inline src/ag-grid-styles.scss:2]: This imports the base styles for ag-grid, and then we override some specific class styles.

File: src/app/app.component.html [hunk 3]
```diff
@@ -168,12 +172,12 @@
...
           </ng-container>
 
-          <button mat-button class="d-none d-md-inline nav-button" 
+          <a mat-button class="d-none d-md-inline nav-button" 
             [routerLink]="[signedOnUser.value.userInfo.membershipType, signedOnUser.value.userInfo.membershipId]"          
             *ngIf="(signedOnUser|async)!=null">            
             <fa-icon [icon]="PLATFORMS_DICT[(signedOnUser|async).userInfo.membershipType+''].icon"></fa-icon>
             My Info
-          </button>
+          </a>
```
Comment (asherbig) [inline src/app/app.component.html:180]: gives url preview when it's an anchor tag

File: src/app/app.module.ts [hunk 4]
```diff
@@ -100,7 +108,8 @@ import { PandaGodrollsService } from './service/panda-godrolls.service';
     PlayerStateService,
     PreferredStatService,
     StorageService,
-    WeekService
+    WeekService,
+    LoggedInGuard
```
Comment (asherbig) [inline src/app/app.module.ts:112]: Ended up not using this guard, but it's still a handy guard to have. It checks to see if there's a currently authenticated user.

File: src/app/bungie-icon/bungie-icon.component.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class BungieIconComponent {
+
+  @Input() set icon(path: string) { this.iconUrl = `http://www.bungie.net${path}`; }
+  @Input() iconClass: string = 'icon';
+
+  public iconUrl: string;
+
+}
```
Comment (asherbig) [inline src/app/bungie-icon/bungie-icon.component.ts:17]: Utility class that saved me a ton of repeated code. I stuck it in the base app for re-use elsewhere

File: src/app/clan/clan-milestones/clan-milestones.component.html [hunk 6]
```diff
@@ -76,7 +76,7 @@
                             <td *ngIf="(hiddenClanMilestones|async).indexOf(mileStoneName.key)===-1"
                                 class="clan-player-ms-cell" [attr.data-label]="mileStoneName.name">
                                 <ng-container *ngFor="let char of player.characters">
-                                    <d2c-milestone-check [char]="char" [mileStoneName]="mileStoneName" detailed="false">
+                                    <d2c-milestone-check [milestoneStatus]="char.milestones[mileStoneName.key]" detailed="false">
```
Comment (asherbig) [inline src/app/clan/clan-milestones/clan-milestones.component.html:79]: I refactored the milestone-check component to take the milestoneStatus object that you currently have keyed in the character object under player. The reason being that the way I store my data, each row holds the status for each character, but doesn't have a reference to the character object or the original milestoneName object.

File: src/app/countdown/countdown.module.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  ],
+  providers: [
+    SecondService
+  ],
+  exports: [
+    TimeUntilPipe,
+    CountdownDirective,
+  ]
+})
+export class CountdownModule { }
```
Comment (asherbig) [inline src/app/countdown/countdown.module.ts:24]: One of my favorite novelties I added. This pattern lets us have live countdown timers in-app. I love it. To me it adds more life to the app when you can see a countdown. Also prevents someone from viewing an out-of-date expiration and thinking they have more time than they actually have.

File: src/app/countdown/directives/countdown.directive.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+@Directive({
+  selector: '[expDateToSeconds]'
+})
+export class CountdownDirective extends Destroyable {
+
+  private dateFeed$: Subject<string> = new Subject();
+
+  @Input() set expDateToSeconds(value: string) {
+    this.dateFeed$.next(value);
```
Comment (asherbig) [inline src/app/countdown/directives/countdown.directive.ts:17]: storing this input into a subject lets us switchMap cancel the previous subscription. It's cleaner imo than managing the subscriptions ourselves.

**Comment by dcaslin** [comment]
Todo list from Asherbig!

**Pull request merged**